### PR TITLE
feat: Implemented structured logging (for warnings)

### LINKED ISSUE #412: Structured logging?
In the code of the dev proxy I see a lot of logging done in an unstructured way, which sets a bad example (in my opinion).

Things like:
https://github.com/microsoft/dev-proxy/blob/d2042e8d0ab351c1eef5bbdc8916e8e522323aae/dev-proxy/ProxyCommandHandler.cs#L86-L88

can also be written as, with exactly the same functionality:
```csharp
 _logger.LogError("New Dev Proxy version {versionNumber} is available.\r\nSee {upgradeUrl} for more information.\r\n", newReleaseInfo.Version, "https://aka.ms/devproxy/upgrade"); 
```

This will output the same error, while also showing logging best practices and allowing the extraction of the important details like the new version number and the url to upgrade.

The same can be set on:
https://github.com/microsoft/dev-proxy/blob/d2042e8d0ab351c1eef5bbdc8916e8e522323aae/dev-proxy/ProxyCommandHandler.cs#L95-L106

```csharp
  catch (Exception ex) {
      _logger.LogError(ex, "An error occurred while running Dev Proxy {message}", ex.Message.ToString());
      var inner = ex.InnerException;

      while (inner is not null) {
          _logger.LogError(inner, "============ Inner exception ============\r\n{message}", inner.Message.ToString());
          inner = inner.InnerException;
      }
}
```

What do you guys think? If I would create a pr with some of these logging "issues" fixed is that something that you would accept into this code?

### PR DESCRIPTION
**Pull request by svrooij**
Fixed #412

Not everything is structured logging just yet, but this is a good start and shows a way to migrate slowly.
Pointing everyone in the right direction, with the obsolete message and the remarks on the original interface.

Actually implementing structured logging would be much easier if dependency injection was already used, so I settled for migrating just the warnings to structured logging.

### CODE REVIEW

**Comment by svrooij** [comment]
And @waldekmastykarz before you ask about the `\n` in the log message somewhere. It's replaced in the **ConsoleLogger**

https://github.com/svrooij/m365-developer-proxy/blob/e28c7c1cb3afe3b97f8e51635d23001f9fa9e202/dev-proxy/ConsoleLogger.cs#L397

**Comment by waldekmastykarz** [comment]
Thanks @svrooij! This is a great start. Here are some comments.

> Not everything is structured logging just yet, but this is a good start and shows a way to migrate slowly.
Pointing everyone in the right direction, with the obsolete message and the remarks on the original interface.

I suggest that we change it all at once. We're still pre-v1 which means we can afford breaking changes now. This won't be as easy in the future, so if there are things that we want to change, now's the time. I'd prefer us to change it all in one go and have a consistent codebase.

> Actually implementing structured logging would be much easier if dependency injection was already used, so I settled for migrating just the warnings to structured logging.

What would be the benefit for us of using DI, where we don't have any plans for replacing any of the components on runtime?

> And @waldekmastykarz before you ask about the `\n` in the log message somewhere. It's replaced in the **ConsoleLogger**
> 
> https://github.com/svrooij/m365-developer-proxy/blob/e28c7c1cb3afe3b97f8e51635d23001f9fa9e202/dev-proxy/ConsoleLogger.cs#L397

Won't that mean that for messages that stem from intercepting requests we'll need to do it twice so that the messages are properly formatted in the console and dev tools, and perhaps other sinks that we might want to use in the future?

It'll be interesting to see how moving to structured logging will affect the LogRequest method which is used by proxy and plugin to communicate information about requests.

I'll mark the PR as draft until it's complete so that we won't merge it by accident.

File: dev-proxy/ConsoleLogger.cs
Comment (waldekmastykarz) [inline dev-proxy/ConsoleLogger.cs:77]: Writing debug message to stderr was on purpose, because they're not considered a part of the command output. Let's reverse this change.

File: dev-proxy/ConsoleLogger.cs
Comment (waldekmastykarz) [inline dev-proxy/ConsoleLogger.cs:35]: Since we're not exposing these methods outside of this logger now, it would be clearer to rename them to `WriteInfo`, etc. That way, we don't confuse with the methods that are a part of `ILogger` and are merely util methods for this particular sink.

File: dev-proxy/ConsoleLogger.cs
Comment (waldekmastykarz) [inline dev-proxy/ConsoleLogger.cs:371]: This is confusing and at the first sight gives the impression that debug logging is always enabled which isn't the case. How could we clarify this?

File: dev-proxy-abstractions/ILogger.cs
Comment (waldekmastykarz) [inline dev-proxy-abstractions/ILogger.cs:48]: Let's remove these remarks. Since we're the sole user of the interface, we don't need them.

**Review by waldekmastykarz — CHANGES_REQUESTED** [review]
Nice! Let's do a couple of changes and clarify some things before we proceed.

File: dev-proxy-plugins/RequestLogs/MinimalPermissionsGuidancePlugin.cs
Comment (svrooij) [reply dev-proxy-plugins/RequestLogs/MinimalPermissionsGuidancePlugin.cs:197]: How do I test this? Didn't know you can even see the data in Chrome DevTools

File: dev-proxy-abstractions/ILogger.cs
Comment (svrooij) [reply dev-proxy-abstractions/ILogger.cs:54]: Yes!

File: dev-proxy/ConsoleLogger.cs
Comment (svrooij) [reply dev-proxy/ConsoleLogger.cs:347]: I checked the source from the official [ConsoleLogger](https://github.com/dotnet/runtime/blob/cd850233414468a8ffe84fc8fa05a36d0b3b1316/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs#L47-L50) which does that also. So yes, why not.

File: dev-proxy/ConsoleLogger.cs
Comment (svrooij) [reply dev-proxy/ConsoleLogger.cs:77]: Changed back

File: dev-proxy/ConsoleLogger.cs
Comment (svrooij) [reply dev-proxy/ConsoleLogger.cs:35]: Yes

File: dev-proxy/ConsoleLogger.cs
Comment (svrooij) [reply dev-proxy/ConsoleLogger.cs:371]: I renamed the property to CurrentLevel, and added comment

File: dev-proxy-abstractions/ILogger.cs
Comment (svrooij) [reply dev-proxy-abstractions/ILogger.cs:48]: What about others building plugins?

File: dev-proxy-abstractions/ILogger.cs
Comment (svrooij) [reply dev-proxy-abstractions/ILogger.cs:56]: Because with structured logging the whether or not a specific level is enabled usually comes from a runtime dynamic configuration. But for this proxy which only partially uses structured logging it makes sense to change the level on the logger itself.

Comment (waldekmastykarz) [reply dev-proxy-abstractions/ILogger.cs]: This still doesn't explain why we need a method to change it and can't use the existing property. What am I missing?

File: dev-proxy-plugins/RequestLogs/MinimalPermissionsGuidancePlugin.cs
Comment (waldekmastykarz) [reply dev-proxy-plugins/RequestLogs/MinimalPermissionsGuidancePlugin.cs]: https://learn.microsoft.com/en-us/microsoft-cloud/dev/dev-proxy/how-to/use-dev-proxy-with-dotnet-docker?pivots=client-operating-system-windows

File: dev-proxy-abstractions/ILogger.cs
Comment (waldekmastykarz) [reply dev-proxy-abstractions/ILogger.cs]: We're changing the interface in a backwards incompatible way which will require anyone who built a plugin in the past to adapt. That said: the new way of logging is the only way we're exposing and our plugins, and in the future docs, will give enough guidance.

File: dev-proxy-abstractions/ILogger.cs
Comment (svrooij) [reply dev-proxy-abstractions/ILogger.cs]: What do you think of removing the LogLevel from the dev proxy and just use the LogLevel from  Microsoft.Extensions.Logging? How does that work when loading it from settings?

Comment (svrooij) [reply dev-proxy-abstractions/ILogger.cs]: This [code](https://github.com/microsoft/dev-proxy/blob/f64d765a3f0099941942ae9412c74ba5ebe6848a/dev-proxy/ProxyConfiguration.cs#L32-L33) parses `debug`, `info`, `warn` and `error`. That will no longer work, does that then require a modification in the docs? Or do we keep the old enum to set the value and convert it in `SetLogLevel(..)` where it's converted to `Microsoft.Extensions.Logging.LogLevel`?

Comment (waldekmastykarz) [reply dev-proxy-abstractions/ILogger.cs:46]: I'm okay with aligning with `Microsoft.Extensions.Logging.LogLevel`. When we merge this PR, we'll update docs accordingly.

File: dev-proxy-plugins/RequestLogs/MinimalPermissionsGuidancePlugin.cs
Comment (svrooij) [reply dev-proxy-plugins/RequestLogs/MinimalPermissionsGuidancePlugin.cs]: I've tested the output in docker and in the regular terminal.

File: dev-proxy-abstractions/ILogger.cs
Comment (svrooij) [reply dev-proxy-abstractions/ILogger.cs]: removed

File: dev-proxy/ConsoleLogger.cs
Comment (waldekmastykarz) [inline dev-proxy/ConsoleLogger.cs:363]: Is there a reason we need to define a private member instead of `public LogLevel LogLevel { get; set; }`?

File: dev-proxy/ProxyCommandHandler.cs
Comment (svrooij) [reply dev-proxy/ProxyCommandHandler.cs]: comments removed

**Review by waldekmastykarz — APPROVED** [review]

**Pull request merged**
